README.md
# XHR As Promised

Yet another XMLHttpRequest wrapper that combines some of the very good but painfully tiny components (rangermauve/xhr, rangermauve/xhr-promise, yields/xhr, matthewp/xhrerror) into a single module for easier consumption. It also adds a convenience wrapper for doing JSON service calls.

## Where to get it?

XHR is available as a [giblet](https://github.com/evs-chris/xhr-as-promised), a [component](https://github.com/componentjs/component), and a pre-assembled UMD module.

All of the pre-built files live in tags on the build branch.

## Usage

XHR provides a library init function that you can supply options to get back an XHR factory function. Notably, if the environment you're using doesn't already supply a global Promise, you have to provide one.

#### `XHRAsPromised(options)`
This is the default export library init function. If is used to set up a request factory function.

##### `options`
* `promise` - the promise constructor to use. It is expected to supply at least a `Promise.reject` function and to turn out usable `then`ables. If none is provided, then the global Promise will be used. If no global can be found, an `Error` will be thrown.
* `onbegin(request)` - a callback function that will be called synchronously just before *any* request is opened
* `onend(request)` - a callback function that will be called synchronously just before *any* request promise is resolved, even if it is a failure

#### `xhr(options)`
This is the function that you will use to make actual requests. It takes an options object or a URL. It also has helper functions for simple requests and a helper object for simple JSON requests. This and all its direct helper functions return a Promise for a completed request.

##### `options`
*All of the options XMLHttpRequest supports*
* `url` - the URL to request
* `method` - the method to use - defaults to `GET`
* `headers` - object - header key/value pairs to set with `setRequestHeader`
* `onbegin(request)` - a callback function that will be called synchronously just before *this* request is opened
* `onend(request)` - a callback function that will be called synchronously just before *this* request promise is resolved, even if it is a failure
* `binary` - boolean - use `sendAsBinary` and if it's not available, return a rejected Promise
* `type` - string - what to send as the Content-Type; defaults to `application/json` for the JSON helpers and `application/x-www-form-uriencoded` for everything else
* `query` - object - key/value pairs to turn into a query string and insert into the url before connecting

##### Helpers
* `xhr.get(url, [options])`
* `xhr.post(url, data, [options])`
* `xhr.put(url, data, [options])`
* `xhr.delete(url, data, [options])` - alias `xhr.del(url, data)` - thanks IE8

##### JSON Helpers
These functions all set the `Accept` header to `application/json` and return a Promise that resolves to the returned object.

* `xhr.json(options)` - JSON version of `xhr(options)`
* `xhr.json.get(url, [options])`
* `xhr.json.post(url, data, [options])`
* `xhr.json.put(url, data, [options])`
* `xhr.json.delete(url, data, [options])` - alias `xhr.json.del(url, data)`

## License
Copyright (c) 2014 Chris Reeves. Released under an [MIT license](https://github.com/evs-chris/xhr-as-promised/blob/master/LICENSE.md).

Probably (c) some others too (rangermauve, yields, matthewp).


CHANGELOG.md
## 0.3.0

* Make any available response headers available on the error object as `responseHeaders` for failed requests.

## 0.2.1

* __BUG:__ The query string will now only allow booleans, strings, and numbers. Any other values will be skipped.

## 0.2.0

* There is now a `query` request option to make building a query string from an object a little easier.

## 0.1.1

* __BUG:__ Fixes typo that caused an exception trying to come up with a content type for posts that lack headers.

## 0.1.0

* More flexible option handling, including support for `responseType` and options for each helper.
* __BUG?:__ JSON helpers only parse the JSON if the response `Content-Type` contains `'json'`.

## 0.0.4

* __BUG?:__ JSON helpers now expect objects instead of pre-strinified JSON.
* __BUG:__ JSON helpers set the proper content type headers.
* There is now a `type` request option to set the content type more easily.

## 0.0.3

* __BUG:__ Fixes yet another issue caused by late-night git pushes.

## 0.0.2

* __BUG:__ Fixes stupid typo that completely broke receiving responses.

## 0.0.1

Initial version


gobblefile.js
var gobble = require('gobble');
var path = require('path');

var parts = [], dev = gobble.env() !== 'production';

var transpilerOptions = {
  blacklist: ['modules']
};

// in dev mode, make libraries and the sandbox available
if (dev) {
  parts.push(gobble('sandbox').moveTo('sandbox'));
  parts.push(gobble('giblets').transform('giblets').moveTo('lib'));
}

var source = gobble('src');

var es5source = gobble(source).transform('6to5', transpilerOptions);

// build the main UMD module
var moduled = gobble(es5source)
  .transform('esperanto-bundle', {
    entry: 'index.js',
    dest: 'xhr-as-promised.js',
    type: 'umd',
    name: 'XHRAsPromised',
    strict: true
  });

// for production, build the component too
if (!dev) {
  parts.push(gobble(es5source).transform('esperanto-bundle', {
    entry: 'index.js',
    dest: 'component/index.js',
    type: 'cjs',
    strict: true
  }));
}

parts.push(moduled.
  transform('relative-sourcemaps', { prefix: 'xhr-as-promised' })
);

module.exports = gobble(parts);


component/index.js
'use strict';

"use strict";

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

var index = build;
/* global XMLHttpRequest, ActiveXObject, window, global */

function XHR() {
  if (window.XMLHttpRequest) {
    return new XMLHttpRequest();
  }try {
    return new ActiveXObject("msxml2.xmlhttp.6.0");
  } catch (e) {}
  try {
    return new ActiveXObject("msxml2.xmlhttp.3.0");
  } catch (e) {}
  try {
    return new ActiveXObject("msxml2.xmlhttp");
  } catch (e) {}
}

function contains(array, value) {
  var i = array.length;
  while (i--) if (array[i] === value) {
    return true;
  }return false;
}

var okStatus = [200, 304, 0];
function requestOk(req) {
  return contains(okStatus, req.status);
}

var XHRError = (function (Error) {
  function XHRError(message, status, request) {
    _classCallCheck(this, XHRError);

    this.message = message;
    this.status = status;
    this.request = request;
    this.responseHeaders = request.responseHeaders;
  }

  _inherits(XHRError, Error);

  return XHRError;
})(Error);

function callbacks(name, req, targets) {
  for (var i = 0; i < targets.length; i++) {
    if (typeof targets[i][name] === "function") targets[i][name](req);
  }
}

function merge() {
  var res = {},
      arg,
      k;
  for (var i = 0; i < arguments.length; i++) {
    arg = arguments[i];
    if (typeof arg === "object") {
      for (k in arg) {
        res[k] = arg[k];
      }
    }
  }
  return res;
}

var betweenQueryAndHash = /\?[^#]*#/;
var afterQuery = /\?.*/;function build(opts) {
  var Promise = opts.promise || opts.Promise || (window || {}).Promise || (global || {}).Promise;

  if (!Promise) throw new Error("I really need a Promise");

  function xhr() {
    var options = arguments[0] === undefined ? {} : arguments[0];
    var req = XHR();

    options.headers = options.headers || {};

    if (options.binary && !req.sendAsBinary) {
      return Promise.reject(new Error("This browser does not support binary XHRs."));
    }callbacks("onbegin", req, [opts, options]);

    if (typeof options === "string") options = { url: options };

    if (options.query && typeof options.query === "object") {
      var qs = "";
      for (var k in options.query) {
        var v = options.query[k];
        if (typeof v === "string" || typeof v === "number" || typeof v === "boolean") {
          qs += (qs ? "&" : "") + ("" + encodeURIComponent(k) + "=" + encodeURIComponent(v));
        }
      }

      var hash = options.url.indexOf("#") !== -1,
          query = options.url.indexOf("?") !== -1;

      if (hash && query) {
        options.url = options.url.replace(betweenQueryAndHash, "?" + qs + "#");
      } else if (hash) {
        options.url = options.url.replace("#", "?" + qs + "#");
      } else if (query) {
        options.url = options.url.replace(afterQuery, "?" + qs);
      } else {
        options.url += "?" + qs;
      }
    }

    req.open(options.method || "GET", options.url, true);
    if (options.credentials) req.withCredentials = true;

    for (var k in options.headers) {
      req.setRequestHeader(k, options.headers[k]);
    }if (options.method === "POST" && (!("Content-Type" in options.headers) && !("type" in options))) {
      req.setRequestHeader("Content-Type", "application/x-www-form-urlencoded");
    } else if ("type" in options) {
      req.setRequestHeader("Content-Type", options.type);
    }

    if ("responseType" in options) req.responseType = options.responseType;

    var res = new Promise(function (ok, fail) {
      req.onreadystatechange = function () {
        if (req.readyState !== 4) return;

        callbacks("onend", req, [opts, options]);

        if (requestOk(req)) {
          ok(req);
        } else {
          var _res = req.responseHeaders = {};
          if (typeof req.getAllResponseHeaders === "function") {
            var headers = req.getAllResponseHeaders().split("\n");
            for (var i = 0; i < headers.length; i++) {
              var n = headers[i].substr(0, headers[i].indexOf(":"));
              if (!n) continue;
              var v = headers[i].substr(n.length + 1);
              _res[n] = v;
              _res[n.toLowerCase()] = v;
            }
          }
          fail(new XHRError("Server responded with a status of " + req.status, req.status, req));
        }
      };
    });

    req[options.binary ? "sendAsBinary" : "send"](options.data || void 0);

    return res;
  }

  (function () {
    xhr.get = function get(url, opts) {
      return xhr(merge({ url: url }, opts));
    };
    xhr.post = function post(url, data, opts) {
      return xhr(merge({ method: "POST", url: url, data: data }, opts));
    };
    xhr.put = function put(url, data, opts) {
      return xhr(merge({ method: "PUT", url: url, data: data }, opts));
    };
    xhr["delete"] = xhr.del = function del(url, data, opts) {
      return xhr(merge({ method: "DELETE", url: url, data: data }, opts));
    };
  })();

  (function () {
    var json = xhr.json = function json() {
      var options = arguments[0] === undefined ? {} : arguments[0];
      if (typeof options === "string") options = { url: options };
      var headers = options.headers = options.headers || {};
      if (!("Accept" in headers)) headers.Accept = "application/json,*/*";
      if (!("Content-Type" in headers) && !("type" in options)) headers["Content-Type"] = "application/json";
      if (typeof options.data !== "string") options.data = JSON.stringify(options.data || "");
      return xhr(options).then(function (res) {
        if ((res.getResponseHeader("Content-Type") || "").toLowerCase().indexOf("json") !== -1) {
          return JSON.parse(res.responseText);
        } else return res;
      });
    };
    json.get = function get(url, opts) {
      return json(merge({ url: url }, opts));
    };
    json.post = function post(url, data, opts) {
      return json(merge({ method: "POST", url: url, data: data }, opts));
    };
    json.put = function put(url, data, opts) {
      return json(merge({ method: "PUT", url: url, data: data }, opts));
    };
    json["delete"] = json.del = function del(url, data, opts) {
      return json(merge({ method: "DELETE", url: url, data: data }, opts));
    };
  })();

  return xhr;
}

exports['default'] = index;
//# sourceMappingURL=index.js.map


xhr-as-promised.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  factory((global.XHRAsPromised = {}))
}(this, function (exports) { 'use strict';

  "use strict";

  var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

  var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

  var index = build;
  /* global XMLHttpRequest, ActiveXObject, window, global */

  function XHR() {
    if (window.XMLHttpRequest) {
      return new XMLHttpRequest();
    }try {
      return new ActiveXObject("msxml2.xmlhttp.6.0");
    } catch (e) {}
    try {
      return new ActiveXObject("msxml2.xmlhttp.3.0");
    } catch (e) {}
    try {
      return new ActiveXObject("msxml2.xmlhttp");
    } catch (e) {}
  }

  function contains(array, value) {
    var i = array.length;
    while (i--) if (array[i] === value) {
      return true;
    }return false;
  }

  var okStatus = [200, 304, 0];
  function requestOk(req) {
    return contains(okStatus, req.status);
  }

  var XHRError = (function (Error) {
    function XHRError(message, status, request) {
      _classCallCheck(this, XHRError);

      this.message = message;
      this.status = status;
      this.request = request;
      this.responseHeaders = request.responseHeaders;
    }

    _inherits(XHRError, Error);

    return XHRError;
  })(Error);

  function callbacks(name, req, targets) {
    for (var i = 0; i < targets.length; i++) {
      if (typeof targets[i][name] === "function") targets[i][name](req);
    }
  }

  function merge() {
    var res = {},
        arg,
        k;
    for (var i = 0; i < arguments.length; i++) {
      arg = arguments[i];
      if (typeof arg === "object") {
        for (k in arg) {
          res[k] = arg[k];
        }
      }
    }
    return res;
  }

  var betweenQueryAndHash = /\?[^#]*#/;
  var afterQuery = /\?.*/;function build(opts) {
    var Promise = opts.promise || opts.Promise || (window || {}).Promise || (global || {}).Promise;

    if (!Promise) throw new Error("I really need a Promise");

    function xhr() {
      var options = arguments[0] === undefined ? {} : arguments[0];
      var req = XHR();

      options.headers = options.headers || {};

      if (options.binary && !req.sendAsBinary) {
        return Promise.reject(new Error("This browser does not support binary XHRs."));
      }callbacks("onbegin", req, [opts, options]);

      if (typeof options === "string") options = { url: options };

      if (options.query && typeof options.query === "object") {
        var qs = "";
        for (var k in options.query) {
          var v = options.query[k];
          if (typeof v === "string" || typeof v === "number" || typeof v === "boolean") {
            qs += (qs ? "&" : "") + ("" + encodeURIComponent(k) + "=" + encodeURIComponent(v));
          }
        }

        var hash = options.url.indexOf("#") !== -1,
            query = options.url.indexOf("?") !== -1;

        if (hash && query) {
          options.url = options.url.replace(betweenQueryAndHash, "?" + qs + "#");
        } else if (hash) {
          options.url = options.url.replace("#", "?" + qs + "#");
        } else if (query) {
          options.url = options.url.replace(afterQuery, "?" + qs);
        } else {
          options.url += "?" + qs;
        }
      }

      req.open(options.method || "GET", options.url, true);
      if (options.credentials) req.withCredentials = true;

      for (var k in options.headers) {
        req.setRequestHeader(k, options.headers[k]);
      }if (options.method === "POST" && (!("Content-Type" in options.headers) && !("type" in options))) {
        req.setRequestHeader("Content-Type", "application/x-www-form-urlencoded");
      } else if ("type" in options) {
        req.setRequestHeader("Content-Type", options.type);
      }

      if ("responseType" in options) req.responseType = options.responseType;

      var res = new Promise(function (ok, fail) {
        req.onreadystatechange = function () {
          if (req.readyState !== 4) return;

          callbacks("onend", req, [opts, options]);

          if (requestOk(req)) {
            ok(req);
          } else {
            var _res = req.responseHeaders = {};
            if (typeof req.getAllResponseHeaders === "function") {
              var headers = req.getAllResponseHeaders().split("\n");
              for (var i = 0; i < headers.length; i++) {
                var n = headers[i].substr(0, headers[i].indexOf(":"));
                if (!n) continue;
                var v = headers[i].substr(n.length + 1);
                _res[n] = v;
                _res[n.toLowerCase()] = v;
              }
            }
            fail(new XHRError("Server responded with a status of " + req.status, req.status, req));
          }
        };
      });

      req[options.binary ? "sendAsBinary" : "send"](options.data || void 0);

      return res;
    }

    (function () {
      xhr.get = function get(url, opts) {
        return xhr(merge({ url: url }, opts));
      };
      xhr.post = function post(url, data, opts) {
        return xhr(merge({ method: "POST", url: url, data: data }, opts));
      };
      xhr.put = function put(url, data, opts) {
        return xhr(merge({ method: "PUT", url: url, data: data }, opts));
      };
      xhr["delete"] = xhr.del = function del(url, data, opts) {
        return xhr(merge({ method: "DELETE", url: url, data: data }, opts));
      };
    })();

    (function () {
      var json = xhr.json = function json() {
        var options = arguments[0] === undefined ? {} : arguments[0];
        if (typeof options === "string") options = { url: options };
        var headers = options.headers = options.headers || {};
        if (!("Accept" in headers)) headers.Accept = "application/json,*/*";
        if (!("Content-Type" in headers) && !("type" in options)) headers["Content-Type"] = "application/json";
        if (typeof options.data !== "string") options.data = JSON.stringify(options.data || "");
        return xhr(options).then(function (res) {
          if ((res.getResponseHeader("Content-Type") || "").toLowerCase().indexOf("json") !== -1) {
            return JSON.parse(res.responseText);
          } else return res;
        });
      };
      json.get = function get(url, opts) {
        return json(merge({ url: url }, opts));
      };
      json.post = function post(url, data, opts) {
        return json(merge({ method: "POST", url: url, data: data }, opts));
      };
      json.put = function put(url, data, opts) {
        return json(merge({ method: "PUT", url: url, data: data }, opts));
      };
      json["delete"] = json.del = function del(url, data, opts) {
        return json(merge({ method: "DELETE", url: url, data: data }, opts));
      };
    })();

    return xhr;
  }

  exports['default'] = index;

}));
//# sourceMappingURL=xhr-as-promised.js.map
